src/ivalue_manager.rs

```diff
@@ -0,0 +1,546 @@
+            } else {
+                Err(err_json(&v, "array"))
+            }
+        })?;
+        match res {
+            None => Err(RedisError::String(err_msg_json_path_doesnt_exist())),
+            Some(l) => Ok(l),
+        }
+    }
+
+    fn clear(&mut self, path: Vec<String>) -> Result<usize, RedisError> {
+        let mut cleared = 0;
+        self.do_op(path, |mut v| match v.type_() {
+            ValueType::Object => {
+                let obj = v.as_object_mut().unwrap();
+                obj.clear();
+                cleared += 1;
+                Ok(Some(v))
+            }
+            ValueType::Array => {
+                let arr = v.as_array_mut().unwrap();
+                arr.clear();
+                cleared += 1;
+                Ok(Some(v))
+            }
+            _ => Ok(Some(v)),
+        })?;
+        Ok(cleared)
+    }
+}
+
+pub struct IValueKeyHolderRead {
+    key: RedisKey,
+}
+
+impl ReadHolder<IValue> for IValueKeyHolderRead {
+    fn get_value(&self) -> Result<Option<&IValue>, RedisError> {
+        let key_value = self.key.get_value::<RedisJSON<IValue>>(&REDIS_JSON_TYPE)?;
+        match key_value {
+            Some(v) => Ok(Some(&v.data)),
+            None => Ok(None),
+        }
+    }
+}
+
+pub struct RedisIValueJsonKeyManager<'a> {
+    pub phantom: PhantomData<&'a u64>,
+}
+
+impl<'a> Manager for RedisIValueJsonKeyManager<'a> {
+    type WriteHolder = IValueKeyHolderWrite<'a>;
+    type ReadHolder = IValueKeyHolderRead;
+    type V = IValue;
+    type O = IValue;
+
+    fn open_key_read(
+        &self,
+        ctx: &Context,
+        key: &RedisString,
+    ) -> Result<IValueKeyHolderRead, RedisError> {
+        let key = ctx.open_key(key);
+        Ok(IValueKeyHolderRead { key })
+    }
+
+    fn open_key_write(
+        &self,
+        ctx: &Context,
+        key: RedisString,
+    ) -> Result<IValueKeyHolderWrite<'a>, RedisError> {
+        let key_ptr = ctx.open_key_writable(&key);
+        Ok(IValueKeyHolderWrite {
+            key: key_ptr,
+            key_name: key,
+            val: None,
+        })
+    }
+
+    fn from_str(&self, val: &str, format: Format) -> Result<Self::O, Error> {
+        match format {
+            Format::JSON => Ok(serde_json::from_str(val)?),
+            Format::BSON => decode_document(&mut Cursor::new(val.as_bytes()))
+                .map(|docs| {
+                    let v = if !docs.is_empty() {
+                        docs.iter().next().map_or_else(
+                            || IValue::NULL,
+                            |(_, b)| {
+                                let v: serde_json::Value = b.clone().into();
+                                let mut out = serde_json::Serializer::new(Vec::new());
+                                v.serialize(&mut out).unwrap();
+                                self.from_str(
+                                    &String::from_utf8(out.into_inner()).unwrap(),
+                                    Format::JSON,
```

Comment:
Not for now (there are missing implementations and tests for BSON), but maybe could be done without passing through JSON string?

Reply:
We need to add support for this on ijson repository